pkg/kubectl/cmd/set/set_volume.go

```diff
@@ -0,0 +1,958 @@
+			storageAnnClass: v.ClaimClass,
+		}
+	}
+	return pvc
+}
+
+func (o *VolumeOptions) setVolumeSource(kv *api.Volume) error {
+	var err error
+	opts := o.AddOpts
+	if len(opts.Type) > 0 {
+		err = setVolumeSourceByType(kv, opts)
+	} else if len(opts.Source) > 0 {
+		err = json.Unmarshal([]byte(opts.Source), &kv.VolumeSource)
+	}
+	return err
+}
+
+func (o *VolumeOptions) setVolumeMount(spec *api.PodSpec, info *resource.Info) error {
+	opts := o.AddOpts
+	containers, _ := selectContainers(spec.Containers, o.Containers)
+	if len(containers) == 0 && o.Containers != "*" {
+		fmt.Fprintf(o.Err, "warning: %s/%s does not have any containers matching %q\n", info.Mapping.Resource, info.Name, o.Containers)
+		return nil
+	}
+
+	for _, c := range containers {
+		for _, m := range c.VolumeMounts {
+			if path.Clean(m.MountPath) == path.Clean(opts.MountPath) && m.Name != o.Name {
+				return fmt.Errorf("volume mount '%s' already exists for container '%s'", opts.MountPath, c.Name)
+			}
+		}
+		for i, m := range c.VolumeMounts {
+			if m.Name == o.Name {
+				c.VolumeMounts = append(c.VolumeMounts[:i], c.VolumeMounts[i+1:]...)
+				break
+			}
+		}
+		volumeMount := &api.VolumeMount{
+			Name:      o.Name,
+			MountPath: path.Clean(opts.MountPath),
+		}
+		if len(opts.SubPath) > 0 {
+			volumeMount.SubPath = path.Clean(opts.SubPath)
+		}
+		c.VolumeMounts = append(c.VolumeMounts, *volumeMount)
+	}
+	return nil
+}
+
+func (o *VolumeOptions) getVolumeName(spec *api.PodSpec, singleResource bool) (string, error) {
+	opts := o.AddOpts
+	if opts.Overwrite {
+		// Multiple resources can have same mount-path for different volumes,
+		// so restrict it for single resource to uniquely find the volume
+		if !singleResource {
+			return "", fmt.Errorf("you must specify --name for the volume name when dealing with multiple resources")
+		}
+		if len(opts.MountPath) > 0 {
+			containers, _ := selectContainers(spec.Containers, o.Containers)
+			var name string
+			matchCount := 0
+			for _, c := range containers {
+				for _, m := range c.VolumeMounts {
+					if path.Clean(m.MountPath) == path.Clean(opts.MountPath) {
+						name = m.Name
+						matchCount += 1
+						break
+					}
+				}
+			}
+
+			switch matchCount {
+			case 0:
+				return "", fmt.Errorf("unable to find the volume for mount-path: %s", opts.MountPath)
+			case 1:
+				return name, nil
+			default:
+				return "", fmt.Errorf("found multiple volumes with same mount-path: %s", opts.MountPath)
+			}
+		} else {
+			return "", fmt.Errorf("ambiguous --overwrite, specify --name or --mount-path")
+		}
+	} else { // Generate volume name
+		name := names.SimpleNameGenerator.GenerateName(volumePrefix)
+		if len(o.Output) == 0 {
+			fmt.Fprintf(o.Err, "info: Generated volume name: %s\n", name)
+		}
+		return name, nil
+	}
+}
+
+func (o *VolumeOptions) addVolumeToSpec(spec *api.PodSpec, info *resource.Info, singleResource bool) error {
+	opts := o.AddOpts
+	if len(o.Name) == 0 {
+		var err error
```

Comment:
Better change to

```
if o.Name, err := o.getVolumeName(spec, singleResource); err != nil {
	return err
}
```

Reply:
I think cause o.Name is an existed variable, so your way is not suitable

Reply:
```
var err error
if o.Name, err = o.getVolumeName(spec, singleResource); err != nil {
	return err
}

```
did you mean this?

Reply:
SGTM